### ISSUE TITLE
how to specify padding?

### ISSUE THREAD

**Issue by etienne87** [facebookresearch/SparseConvNet#13]
i don't see where to specify padding for convolutions? 
when i create a dummy model with one layer, filter_size=3, filter_stride=1 and compute input_spatial_size for [10,10] output i get 12, 12, which seems to indicate padding = 0.

**Comment by btgraham** [comment]
Submanifold convolutions are automatically padded with padding (size-1)/2. i.e. 3x3 filters have padding of 1. They are used for example in residual blocks where preserving the scale/sparsity is important.

Regular sparse convolutions are not padded. They are generally used for downsampling, ie size=3,stride=2 or size=2,stride=2. These don't have a padding option. Padding could be added as an option, but I am not sure I see a use case. 

I recommend using one of the network builder functions to build your network. See for example: https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/networkArchitectures.py#L21
and 
https://github.com/facebookresearch/SparseConvNet/blob/master/examples/Assamese_handwriting/VGGplus.py

Regards
Ben

**Comment by etienne87** [comment]
Ok! Thank you for answering so quickly. Well, I was thinking just loading weights of standard VGG into its sparse version and finetuning.

**Comment by zjhthu** [comment]
In my use case, my input size are even numbers ,for example 240,  I want to use the [`SparseResNet`](https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/networkArchitectures.py#L148)  builder functions for down-sampling,  in the residual block, filter size is 3, stride is 2, as sparse convolutions dose not support padding,  code will fail in [this assert](https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/convolution.py#L106).
So, could you give me some advice about usage, or could you add the padding option for sparse convolution.

**Comment by zjhthu** [comment]
@btgraham, could you please give me some advice about adding the padding option, if I want to use standard 3x3 convolution.

**Comment by btgraham** [comment]
Input size is not so important for SparseConvNet as for dense ConvNets as extra size has no cost. If your inputs have a maximum size of 240x240, and you want the output to have spatial size 1x1, for feeding to a softmax classifier, you could: 

1) Use 5 levels of size-3 stride-2 downsampling, followed by 7x7 average pooling
------> Input size = 255
2) Use 5 levels of size-3 stride-2 downsampling, followed by a 7x7 convolution
------> Input size = 255
3) Use 6 levels of size-3 stride-2 downsampling, followed by a 4x4 convolution
-------> input size = 319
etc, etc

You can place the 240x240 objects in the large space either 
A) in the centre or 
B) at random (for a bit of data augmentation).

**Comment by zjhthu** [comment]
I got it, thanks for your suggestion. But I may still need the padding option in my later application. I need to get an output which has different structure with input, so regular sparse convolutions are used. Padding is necessary for keeping size.

**Comment by btgraham** [comment]
Can you explain a little bit more about what you are trying to do?

**Comment by zjhthu** [comment]
For my application, input and output have different sparsity structure,  I am going to use Regular Sparse Convolution to predict the output structure.
I  have looked through your amazing project, and I plan to add padding option support for Regular Convolution layer, I think there are mainly three places need to be modified:
1.  add padding option in [`OutputRegionCalculator`](https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/SCN/generic/Geometry/RectangularRegions.h#L111)
2. Modify related interface in  [`convolution.py`](https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/convolution.py), [`Convolution.cu`](https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/SCN/generic/GPU/Convolution.cu) and [`ConvolutionRules.h`](https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/SCN/generic/Geometry/ConvolutionRules.h)
3. Modify [`output.spatial_size`](https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/convolution.py#L106) and remove `assert`
In this way,  the [`input_spatial_size`](https://github.com/facebookresearch/SparseConvNet/blob/master/PyTorch/sparseconvnet/convolution.py#L135) function is unnecessary.
Am I right?

**Comment by btgraham** [comment]
I am still not sure padding is ever needed.
You can just make the input layer size 10**6 by 10**6, put your sparse input in the middle. You are then so far from the boundary that it doesn't matter.
What do you mean by predict the output structure?

**Comment by zjhthu** [comment]
I think you are right. My previous concern is mainly about keeping feature map size unchanged while using Regular sparse convolution. 
Anyway, thanks very much.
